TST: jax random key gen

nstarman opened the pull request:
Seemed to work locally. Not a good fix, but perhaps a diagnosis.

Fixes #42

patrick-kidger commented:
Oh wow, this really surprises me. How did you find this?

The choice of random key should ideally be a runtime-only thing, and shouldn't affect compilation (when these leaks occur). My first guess is that it's specifically this line that is doing something that JAX doesn't like:

https://github.com/patrick-kidger/equinox/blob/f8df0e648bf3af8c9b041fc76f58fcc5b4ab37c6/equinox/internal/_getkey.py#L42

Where as in your example here the key is generated on-demand rather than being stored somewhere.

nstarman commented:
> Oh wow, this really surprises me. How did you find this?

There didn't seem to be rhyme or reason to which test failed, so I took a wild guess! 😆 

> My first guess is that it's specifically this line that is doing something that JAX doesn't like:

So what's next step? I can try making a new GetKey here in quax, or use numpy to make a random seed, or option 3...

nstarman commented:
I've diagnosed further. `jr.PRNGKey` works where `jr.key()` does not! I saw there's some Issue in diffrax about key generation. Is this related?

nstarman commented:
An "easy" refactor I can do is to have the fixture return a key and just split the key in the tests.

```python
@fixture(scope=session)
def seed(): return random.randomint(0, 2**32-1)

@fixture
def getkey(request, seed):
    key = jr.key(seed)
    key = jr.fold_in(key, hash(request.function))
    return key

def test_x(getkey):
    key, *subkeys = jr.split(key, 5)

```

patrick-kidger commented:
I think I'd quite like to know why the original implementation raises an error in the first place! Are you able to isolate a MWE?
If it's flaky in CI then maybe it depends on Python version / on the order in which tests are ran / etc.

nstarman commented:
> I think I'd quite like to know why the original implementation raises an error in the first place!

I suspect you're right that it's the statefulness of `GetKey`, but I can't confirm. 

> Are you able to isolate a MWE?

Not as of yet! It doesn't appear to depend on Python version, or test order (I've tried running specific files).
And now I can't even get it to error, which I guess is a good thing!

nstarman commented:
So I think this PR can be closed soon. I'll make it a draft while the other PRs are in, then close it if there's no further problems.

nstarman commented:
I think I figured it out!
LoRa is the only class to store the `key` and `test_lora.test_materialise` is the only test to check a `pytest.raises`. If I remove that test I never get the tracer leak (which I managed to start reproducing). If I split the key so there isn't any kind of direct reference to `getkey()` then it never seems to error.
So my hypothesis is that `getkey()` +  `pytest.raises` can lead to the tracer leaks! 
This is supported by `test_named.test_init`  where

```
with pytest.raises(ValueError):
        a = jr.normal(getkey(), (3, 3))
        a = named.NamedArray(a, (Foo, Foo))
```
can error, but if I change `getkey()` to `jr.split(get key((), 2)[0]` then it never errors.

nstarman commented:
LMK if you want me to move some of the config changes to different PRs or if it's fine for them to sneak in here.

patrick-kidger commented:
Oh, now that is super mysterious. Indeed I bet it's some kind of shenanigans from the errors, then. I've seen similar kinds of tracer leaks from things like https://github.com/jaraco/keyring/issues/670#issuecomment-2488935432 when exceptions are erroneously held on to.

Anyway, I think this LGTM! So thank you for tracking this down, and merged!
